File: 994-PrisonCellsAfterNDays/994-PrisonCellsAfterNDays.java
// Last updated: 8/5/2025, 2:56:15 PM
class Solution {
    public int[] prisonAfterNDays(int[] cells, int n) {
        HashSet<String> set = new HashSet<>();
        int count = 0;
        for (int i=0; i<n; i++) {
            int[] nextDay = findNext(cells);
            String key = Arrays.toString(nextDay);
            if(!set.contains(key)) {
                set.add(key);
                count++;
            }
            else {
                break;
            }
            cells = nextDay;
        }

        n%=count;
        for(int i=0; i<n; i++) {
            cells = findNext(cells);
        }
        return cells;

    }

    public int[] findNext(int[] cells) {
        int[] nextCells = new int[cells.length];
        for(int i=1; i<cells.length-1; i++) 
            nextCells[i] = cells[i-1] == cells[i+1] ? 1 : 0;
        return nextCells;
    }
}

File: 994-PrisonCellsAfterNDays/994-PrisonCellsAfterNDays.py
# Last updated: 8/5/2025, 2:56:19 PM
class Solution:
    def prisonAfterNDays(self, cells: List[int], N: int) -> List[int]:
        seen = dict()
        isFound = False
        itr = N
        while itr > 0:
            if not isFound:
                state_key = tuple(cells)
                if state_key in seen:
                    itr %= seen[state_key] - itr
                    isFound = True
                else:
                    seen[state_key] = itr
            if itr > 0:
                cells = self.updateCells(cells)
                itr -= 1    
        return cells
    
    def updateCells(self, cells):
        ret = []
        ret.append(0)
        for ele in range(1, len(cells) - 1):
            ret.append(int(cells[ele - 1] == cells[ele + 1]))
        ret.append(0)
        return ret